File: Data-Structure-I/1-Array/P566.cpp
// Date: Sat, Oct 2, 2021
// 566. Reshape the Matrix (https://leetcode.com/problems/reshape-the-matrix/)

// My idea:
// 4x2
// 00 will be 0*2+0 = 0 -> pattern: row * cols + col
// 01 will be 0*2+1 = 1
// 10 will be 1*2+0 = 2
// 11 will be 1*2+1 = 3
// 20 will be 2*2+0 = 4
// 20 will be 2*2+1 = 5
// 30 will be 3*2+0 = 6
// 31 will be 3*2+1 = 7
// 1x4
// 0123
// 4567
// 12 will find out by 1 = 6 / 4 and 2 = 6 % 4 -> pattern: row = x / cols and col = x % cols

// time complexity: O(mn)
// space complexity: O(1)
vector<vector<int>> matrixReshape(vector<vector<int>>& mat, int r, int c) {
	int m = mat.size(), n = mat[0].size();
	if (m*n != r*c) 
		return mat;
	vector<vector<int>> ans(r, vector<int>(c));
	for (int i = 0; i < m; i++) {
		for (int j = 0; j < n; j++) {
			int num = i * n + j;
			ans[num/c][num%c] = mat[i][j];
		}
	}
	return ans;
}

File: Programming-Skills-I/P566.java
package programmingSkills1;

import java.util.Arrays;

/**
 * Date: March 20, 2022
 * 566. Reshape the Matrix
 * https://leetcode.com/problems/reshape-the-matrix/
 */

public class P566 {
    // time: O(m * n)
    // space: O(1)
    public int[][] matrixReshape(int[][] mat, int r, int c) {
        int m = mat.length, n = mat[0].length;
        if (m * n != r * c) return mat;
        int[][] ans = new int[r][c];
        for (int row = 0; row < m; row++) {
            for (int col = 0; col < n; col++) {
                int index = row * n + col;
                ans[index / c][index % c] = mat[row][col];
            }
        }
        return ans;
    }

    public static void main(String[] args) {
        P566 obj = new P566();
        System.out.println("Expected: [[1, 2], [3, 4]]\nReceived: " + Arrays.deepToString(obj.matrixReshape(new int[][]{{1}, {2}, {3}, {4}}, 2, 2)));
    }
}
